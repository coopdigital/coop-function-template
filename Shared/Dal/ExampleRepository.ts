import { Container, CosmosClient } from '@azure/cosmos';
import { Service } from 'typedi';
import AppInsights from '@coop/azure/lib/AppInsights';
import { ExampleModel } from '../Models/ExampleModel';

/** Example Cosmos DB Repository */
@Service()
export default class ExampleRepository {
  /** The cosmosClient, this connection should be maintained across instances */
  private cosmosClient: CosmosClient;
  /** The **Database** container */
  private container: Container;

  /** Initialises the *Cosmos DB* and gets the **database** and **container** */
  constructor() {
    this.cosmosClient = new CosmosClient({
      endpoint: process.env.cosmos_endpoint,
      key: process.env.cosmos_key,
    });

    this.container = this.cosmosClient
      .database(process.env.cosmos_database)
      .container(process.env.cosmos_container);
  }

  /** Gets an exampleModel by Id */
  async getExampleById(id: string): Promise<ExampleModel> {
    let exampleModel: any = null;

    // Get product by ID, second field is the partition key, in this
    // example Id but generally wouldn't be ID.
    const { resource: data } = await this.container.item(id, id).read();

    if (data) {
      exampleModel = data as ExampleModel;
    }

    return exampleModel;
  }

  /** Finds records by **Field 1** */
  async find(field1: string): Promise<Array<ExampleModel>> {
    // Use the Parameters to protect against SQL Injection attacks
    const querySpec = {
      query: 'SELECT * FROM p WHERE p.Field1 = @field1',
      parameters: [{ name: '@field1', value: field1 }],
    };

    const { resources: data } = await this.container.items
      .query(querySpec)
      .fetchAll();

    return data.map((record) => {
      return record as ExampleModel;
    });
  }

  /** Updates the exampleModel record in the DB */
  async updateExampleModel(exampleModel: ExampleModel): Promise<void> {
    await this.container.items.upsert(exampleModel);
  }
}
