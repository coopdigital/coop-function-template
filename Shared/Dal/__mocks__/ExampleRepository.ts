import { Container } from "@azure/cosmos";

const mock = {
  getExampleById: jest.fn((id) => {
    return {
      id: 'd8db7f54-7758-4a5a-bfd1-82e205c80a59',
      Field1: 'Test Data',
      Field2: 1234,
    };
  }),
  find: jest.fn((field1) => {
    return [
      {
        id: 'd8db7f54-7758-4a5a-bfd1-82e205c80a59',
        Field1: 'Test Data',
        Field2: 1234,
      },
    ];
  }),
  updateExampleModel: jest.fn()
};
export default mock;
