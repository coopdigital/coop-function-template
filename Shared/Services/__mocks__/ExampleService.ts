export const mockGet = jest.fn((id) => {
  return {
    Field1: 'Test Data',
    Field2: 1234,
  };
});

export const mockUpdate = jest.fn();

const mock = jest.fn().mockImplementation(() => {
  return { get: mockGet, update: mockUpdate };
});
export default mock;
