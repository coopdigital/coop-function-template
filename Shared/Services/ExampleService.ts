import { Service } from 'typedi';
import ExampleRepository from '../Dal/ExampleRepository';
import { ExampleDto } from '../Dtos/ExampleDto';
import { ExampleModel } from '../Models/ExampleModel';

/** Example Service, abstract business logic from the Function */
@Service()
export default class ExampleService {
  /** Initialises the Service */
  constructor(private exampleRepository: ExampleRepository) {}

  /** Gets a example record by id */
  public async get(id: string): Promise<ExampleDto> {
    let response: ExampleDto = null;

    const exampleData = await this.exampleRepository.getExampleById(id);

    if (exampleData) {
      response = {
        Field1: exampleData.Field1,
        Field2: exampleData.Field2,
      };
    }

    return response;
  }

  /** Updates an example */
  public async update(id: string, example: ExampleDto): Promise<void> {
    const exampleModel: ExampleModel = { id, ...example };

    await this.exampleRepository.updateExampleModel(exampleModel);
  }
}
