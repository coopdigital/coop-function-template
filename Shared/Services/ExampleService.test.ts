import ExampleService from "./ExampleService";
import MockDatabaseRespository from  '../Dal/__mocks__/ExampleRepository';
import { ExampleRepository } from "../Dal/ExampleRepository";

describe('ExampleService Tests',()=>{
    beforeEach(()=>{
        MockDatabaseRespository.getExampleById.mockClear();
        MockDatabaseRespository.updateExampleModel.mockClear();
    })
    it('get - Success', async ()=>{
        const exampleService = new ExampleService(MockDatabaseRespository as unknown as ExampleRepository);
        
        const result = await exampleService.get('TEST ID');

        expect(result).toMatchSnapshot();
    })

    it('get - Not found', async ()=>{
        MockDatabaseRespository.getExampleById.mockImplementation(()=>{ 
            return null;
        })
        const exampleService = new ExampleService(MockDatabaseRespository as unknown as ExampleRepository);
        
        const result = await exampleService.get('TEST ID');

        expect(result).toMatchSnapshot();
    })

    it('update - Success', async () => {
        const testData = {
            Field1:'Test Data',
            Field2:1234
        }    

        const exampleService = new ExampleService(MockDatabaseRespository as unknown as ExampleRepository);

        await exampleService.update('Test ID', testData);

        expect(MockDatabaseRespository.updateExampleModel.mock.calls.length).toBe(1);
        expect(MockDatabaseRespository.updateExampleModel.mock.calls[0][0]).toMatchSnapshot();
    })
})