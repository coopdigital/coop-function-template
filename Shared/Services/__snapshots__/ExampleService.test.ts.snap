// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExampleService Tests get - Not found 1`] = `null`;

exports[`ExampleService Tests get - Success 1`] = `
Object {
  "Field1": "Test Data",
  "Field2": 1234,
}
`;

exports[`ExampleService Tests update - Success 1`] = `
Object {
  "Field1": "Test Data",
  "Field2": 1234,
  "id": "Test ID",
}
`;
